flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/source/reader/PollingDynamoDbStreamsShardSplitReader.java

```diff
@@ -106,6 +118,49 @@ public RecordsWithSplitIds<Record> fetch() throws IOException {
         }
 
         long currentTime = System.currentTimeMillis();
+
+        if (splitContext.splitState.isHasShardEndReached()) {
```

Comment:
I'd separate the body of this `if` into a new private method, e.g:
```java
private RecordsWithSplitIds<Record> handleShardEnd(
        DynamoDbStreamsShardSplitWithContext splitContext, long currentTime) { ... }
```

And inside, I'd invert the `MAX_RETRY_ATTEMPTS_FOR_CHILD_SHARDS` check (return early, and no `else` branch required in that case), so we can minimize the endless nested `if`s.